stream-chat-android-client/src/main/java/io/getstream/chat/android/client/setup/InitializationCoordinator.kt

```diff
@@ -28,20 +33,67 @@ public class InitializationCoordinator private constructor() {
 
     private val userDisconnectedListeners: MutableList<(User?) -> Unit> = mutableListOf()
 
+    /**
+     * Listeners which will get invoked when the web socket goes through connection changes.
+     */
+    private val onSocketConnectionStateChangedListener: MutableList<(SocketConnectionState) -> Unit> = mutableListOf()
+
     /**
      * Adds a listener to user disconnection.
      */
     public fun addUserDisconnectedListener(listener: (User?) -> Unit) {
         userDisconnectedListeners.add(listener)
     }
 
+    /**
+     * Adds a listener used to listen to socket connection changes
+     */
+    public fun addSocketConnectionStateListener(listener: (SocketConnectionState) -> Unit) {
+        onSocketConnectionStateChangedListener.add(listener)
+    }
+
     /**
      * Notifies user disconnection
      */
     internal fun userDisconnected(user: User?) {
         userDisconnectedListeners.forEach { function -> function.invoke(user) }
     }
 
+    /**
+     * Notifies when the socket connection state has been changed.
+     *
+     * @param chatEvent The socket state change event. Pass in only the following event types:
+     * [ConnectedEvent], [ConnectingEvent], [DisconnectedEvent].
+     */
+    internal fun socketConnectionStateChanged(chatEvent: ChatEvent) {
+        val socketConnectionState: SocketConnectionState = when (chatEvent) {
+            is ConnectedEvent -> SocketConnectionState.CONNECTED
+            is ConnectingEvent -> SocketConnectionState.CONNECTING
+            is DisconnectedEvent -> SocketConnectionState.DISCONNECTED
+            else -> {
+                StreamLog.w("InitializationCoordinator.socketConnectionStateChanged") {
+                    "Only ChatEvents of types ConnectedEvent, ConnectingEvent and DisconnectedEvent" +
+                        " should be passed as arguments to this method."
+                }
+
+                return
+            }
+        }
+
+        onSocketConnectionStateChangedListener.forEach { function ->
+            function(socketConnectionState)
+        }
+    }
+
+    /**
+     * Used to represent socket connection events.
+     */
+    public enum class SocketConnectionState {
```

Comment:
Do we need a different state for that? Maybe we can reuse `io.getstream.chat.android.client.models.ConnectionState`? 🤔

Reply:
Good suggestion, will change.